Ports Families gamemode to Dynamic roundstart + midround rulesets

tmtmtl30 opened the pull request:
## About The Pull Request

Families is a decent gamemode. It isn't currently an option during dynamic rounds, unlike every other gamemode. Thus, it should be ported to dynamic. The current families code is not very conducive to this plan of action; even if it were, it would involve the duplication of hundreds of lines of code across 2-3 files that would need to be maintained in parallel in order to preserve parity between the dynamic ruleset and the standalone gamemode. This PR refactors the families code significantly in order to port it to dynamic (in the form of a roundstart ruleset and a rarer, shorter midround ruleset), done in two parts. Firstly, a significant amount of code is moved to the family member and space police antagonist datums. Secondly, the code that cannot be moved to the antagonist datums is instead moved to a new, unique "handler" datum. The families gamemode and ruleset instance this "handler" datum, feed it candidates, modify its variables directly after if necessary, and then call its procs. For example, the post_setup() and execute() procs of the families gamemode and dynamic rulesets respectively each call the "handler" proc post_setup_analogue(). This means that all the code is centralized and easier to maintain.

The mechanical impact on the standalone families gamemode is negligible. The families roundstart dynamic ruleset functions nearly identically to the standalone gamemode, barring slightly different criteria for eligibility. The families midround dynamic ruleset is very similar to the roundstart ruleset, though it is rarer and the space police will arrive 30-15 minutes after the ruleset's drafting as opposed to 60-30 minutes after the beginning of the round for the gamemode and roundstart ruleset. The roundstart dynamic ruleset and the midround dynamic ruleset cannot coexist; no round will have both.

Additionally, some bugs have been fixed as part of the refactor. The gamemode no longer iterates through all family members and cops to see if they're dead every time any person or slime dies, due to a misnamed proc. The space police 5 minute warning notification and arrival notification will occur at appropriate times instead of after the space police's arrival. An unrelated deprecated function and check have been removed. Finally, families are now nearly fully functional if they are added to players using the admin antagonist panel. They can invite others to their family and complete objectives, though there will be no space police, wanted levels, or point gain (outside of family members switching sides).

## Why It's Good For The Game

Families is an interesting gamemode, with a greater emphasis on roleplay and trust than any of our other antagonists. It's a shame that it can't pop up in dynamic, where the unique combinations of antagonists would really let it shine. This fixes that.

## Changelog
:cl:
fix: The Families gamemode and antagonist have been refactored and should perform more sanely in some cases.
add: Families can now appear in dynamic rounds.
/:cl:

Closes #49877
?????? #49935 -- whatever happened is still a mystery, but i think i stopped it from freezing everything and the code structure is so different now that whatever set of circumstances that came to pass to cause this issue might be no longer possible.

tmtmtl30 committed:
get back up to date

hell yeah

tmtmtl30 committed:
i think goof did some bullshit

good to have in

tmtmtl30 committed:
stage one

doesn't compile yet

tmtmtl30 committed:
Merge branch 'master' into dynfamilies

tmtmtl30 committed:
step 2

added newer changes and goof's balance requests. still doesn't compile

tmtmtl30 committed:
phase 3

somewhere along the way, wanted code got fucked, and i have to commit to switch branches to debug it better

tmtmtl30 committed:
Merge branch 'master' into dynfamilies

tmtmtl30 committed:
documentation, first dyn rulesets

-deprecates a proc nobody needed
-documents a lot of things, hopefully enough of them
-reorganizes a lot of things to allow for admin-spawned families
-adds dynamic rulesets (unfinished, debug settings left in)
-hopefully finally fixes wanted level displays. all code sucks; ui code sucks especially. why can i put an acid level on a screen element? Fuck you
-implements fully functional jewish qabbalah system to replace families gamemode. discover secret names of god + knowledge of universe structure through study of tanach and talmud, use them to secure the station for israel 2
-introduces gay agenda to unsuspecting stationgoers
-remind me to check out gang tag code for mentions of ssticker.mode etc. if one of my later commits doesn't say i did that

tmtmtl30 committed:
Merge remote-tracking branch 'upstream/master' into dynfamilies

my midround ruleset wasn't working because pre_execute() wasn't getting called. i remembered there was a PR that fixed this; i'm not actually sure this wasn't an intentional element of midround ruleset design, as my ruleset will be apparently literally the only midround ruleset to use that proc. on the off chance that this oddity causes some poor future coder trouble, i'm sorry, and no, neither of us deserve any better.

tmtmtl30 committed:
finishes up rulesets

-rulesets work now
-hopefully cleaned up the debug code for them
-minor adjustment to check_wanted_level() proc to ensure wanted stars update appearance correctly post-space cops
-i resisted the urge to modify one of the systems so that it was more "elegant" or whatever because knowing my luck it'd result in another 10 years of failures and tests
-checked out gang tag code; seems fine. perhaps the only part of families code i haven't gotten my mitts on. hopefully it won't break
-you can now print fully functional geneseed organs from the roundstart medlathe for use in self-augmentation
-consumed my soul and replaced it with a pathetic, empty husk
-implemented social security number skimming operation
-removed herobrine

tmtmtl30 committed:
addressing issues

from the github

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -208,6 +208,71 @@
+	cost = 25
+	requirements = list(101,101,101,90,80,70,30,15,10,10) // more strict than the roundstart ruleset, because it's midround
+	high_population_requirement = 10
+	flags = TRAITOR_RULESET
```
```suggestion
```
TRAITOR_RULESET is for classic secret mode and shouldn't be used on a ruleset like this.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -461,6 +461,53 @@
+	cost = 30
+	requirements = list(101,101,101,80,70,30,15,10,10,10)
+	high_population_requirement = 10
+	flags = TRAITOR_RULESET
```
```suggestion
```

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -461,6 +461,53 @@
+	requirements = list(101,101,101,80,70,30,15,10,10,10)
+	high_population_requirement = 10
+	flags = TRAITOR_RULESET
+	blocking_rules = list(/datum/dynamic_ruleset/midround/families)
```
You got the blocking_rules the wrong way, this should go the midround ruleset with 
`/datum/dynamic_ruleset/roundstart/families`.

tmtmtl30 replied:
that line is already there, i was just copying what the revs rulesets did.

TheChosenEvilOne replied:
well, this just stops the roundstart ruleset from being triggered AFTER the midround has been triggered.

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_roundstart.dm:
```diff
@@ -461,6 +461,53 @@
+	QDEL_NULL(handler)
+	..()
+
+/datum/dynamic_ruleset/roundstart/families/Destroy()
```
rulesets generally don't get deleted, but if you want to you should make Destroy call clean_up on a lower level.

tmtmtl30 replied:
eh, i'll just remove the proc

TheChosenEvilOne commented on code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:
```diff
@@ -208,6 +208,71 @@
+	QDEL_NULL(handler)
+	..()
+
+/datum/dynamic_ruleset/midround/families/Destroy()
```
See the comment on /datum/dynamic_ruleset/roundstart/families/Destroy()

TheChosenEvilOne reviewed (commented):
Just a quick review on the dynamic side of things.

ATH1909 commented:
Why does this PR have so many downthumbs?

evilcyborgmrt commented:
Because nobody likes families

tmtmtl30 commented:
it's a matter of consistency. if families isn't well-designed, then it should be removed or reworked, not kept in a limbo state and shielded from our most populous server.

i had some ideas for switching up how induction packages are spawned + used, to prevent people from spamming them early in the round and bypassing the balance cap, but i didn't want to bikeshed even more than i already am. if you have mechanical suggestions to improve the mode (adjusting point gain, wanted levels, objectives for families / space police), i'd be happy to hear them.

tmtmtl30 committed:
suggestions

cleanin some things up

Iamgoofball commented:
I will be PRing the new families gangs to this PR to avoid merge conflicts

Shadowflame909 commented:
> Why does this PR have so many downthumbs?

because its just organized greytide with no mechanics.

every round is a families round as long as the greytide wears the same jumpsuit and gets a spray can from the autolathe.

SkeletalElite commented:
keep families off bagil please and thank you

Iamgoofball commented:
Stop posting garbage about whether you like or dislike families in this PR. I will be reporting all posts that do that.

optimumtact commented:
Keep it about the specifics of families in dynamic please, if you want to complain about the gamemode we have a forum.

BeeSting12 commented:
How likely is this ruleset to be used?

kriskog reviewed (commented):
Since families has pretty big round impact it should probably be treated as a highlander ruleset.
```
flags = HIGHLANDER_RULESET
```
This makes it a lot less likely to spawn with other highlanders, so nukies, cult, revs.

tmtmtl30 commented:
the roundstart ruleset has a weight of 2, the same as the roundstart revolutionary or roundstart wizard rulesets. the midround ruleset has a weight of 1, the same as the midround wizard ruleset. as these weights are relative to the others in their category (roundstart rulesets only compete with other roundstart rulesets and midround rulests only compete with other midround rulesets), they can't be compared to each other. 

by weight, the roundstart families ruleset composes 2/24ths or ~8.3% of the roundstart rulesets. this is the same as the revolution ruleset if the families ruleset exists, as they have the same weight -- for reference, if the families ruleset was not present, the revolution ruleset would make up 2/22nds (~9.1%) of the roundstart rulesets by weight. the midround families ruleset is significantly rarer, composing 1/31st or ~3.2% of the roundstart rulesets by weight. however, there are additional requirements for rulesets to be drafted -- they must meet a minimum amount of eligible candidates, minimum population, and minimum threat level dependent on the population. additionally, not all rounds will draft a midround ruleset, so it's hard to be more precise.

kriskog commented:
Since dynamic will roll between 1 and 3 roundstart rulesets, the cost is another pretty important variable. At 25 cost it is effectively more likely to run than revs, as it will fit as secondary/tertiary ruleset more often.

tmtmtl30 committed:
more tweaks

goof requested on discord that i ensure 3 families every time, and adding the highlander ruleset flag makes sense because it means the 1-2 families-revs or families-nukies punch is less likely

tmtmtl30 commented:
yeah, that's true (though roundstart families is 30 cost, not 25)

IndieanaJones commented:
How often does the midround variant of this show up?  Do I have to be concerned with families randomly showing up and ending a round due to the forced presence of Space Police?  Is there any way to stop it once the gamemode is triggered to prevent the Space Police?

Also, how is death calculated in a midround situation?  Do the Space Police factor in deaths on the station prior to the arrival of the gangs?  Does it count only crew, or do other carbons also get tallied?

tmtmtl30 commented:
midround families should be pretty rare, but it's hard for me to anchor it to something else since pretty much every other midround ruleset misunderstood the "high_population_requirement" variable, meaning that they're way less likely to spawn above 55 population (the variable represents the amount of threat needed to draft the rule, separate from the rule's _cost_, during a 55 or higher pop round, and the rulesets have incorrectly set it to 50, when the 48-53 pop range requirement is 10). i'll open another PR to fix that. it's maybe a lot less likely than midround nukies, though, and i've never thought of those as particularly common. on populations in the 48-54 range, it's slightly less likely than midround wizard, and <1/4th as likely as a blob or a space dragon.

currently, the midround ruleset would count deaths from the start of the shift, which does seem a little bit silly. i'll fix that too.

tmtmtl30 commented:
on second thought, i'm not too sure. if there are a lot of deaths before the midround ruleset activates, then presumably things aren't going great on-station and leaving quickly is probably in order _and_ a large force of highly-equipped cops would make sense; _then_ you run into the problem of cucking antagonists by bringing in a bunch of highly-equipped soldiers. i'm curious what other people think about this

tmtmtl30 commented:
okay, i've added a check such that if the deaths_during_shift is greater than half of the people who were readied up at roundstart, then the families midround ruleset will fail. this is in addition to the requirement of 36 living players.

tmtmtl30 committed:
adjustments for midround ruleset

-wanted level no longer cares about deaths that happened before the handler's pre_setup_analogue() proc is called
-midround ruleset will not trigger if more people have died during the shift than there were people readied up at roundstart (in addition to the minimum population of 36).
-changes to antagonist code; only nonhumans may roll antagonist. if you have more than 1 hour of time spent playing as a human you are ineligible. something something 13%

nightred commented:
I believe that the deaths count related to bringing in the cops should only count after the game mode has been activated, and should have a way of ending without ending the round completely.

tmtmtl30 commented:
that's the way i have deathcounting handled. in families, they don't end the round, just delay the shuttle's departure until 10 minutes after the cops arrive. i could probably change that too.

tmtmtl30 committed:
#51315 compatibility

removes useless "high_population_requirement" vars on families rulesets

tmtmtl30 committed:
moves stuff

team creation should have gone in create_team() not on_gain() since the former gets called directly before the latter

tmtmtl30 commented:
reeeeee

tmtmtl30 committed:
shuffles some procs

removes unnecessasry blocking_rules list for roundstart families gamemode

tmtmtl30 committed:
Merge remote-tracking branch 'upstream/master' into dynfamilies

merge confliiiicts

kriskog commented:
Ill look at this soonish

kriskog reviewed (commented):
Merge conflict, otherwise seems merge ready.

tmtmtl30 committed:
Merge branch 'master' into dynfamilies

here's hoping this works

tmtmtl30 commented:
should be fixed

kriskog commented:
We might need to tweak stuff, like the chance of occurence. Lets see how it works out
